src/basic/unit-name.c

```diff
@@ -665,6 +665,31 @@ int unit_name_mangle_with_suffix(const char *name, UnitNameMangle flags, const c
         return 0;
 }
 
+bool service_unit_name_is_valid(const char *name) {
+        _cleanup_free_ char *prefix = NULL, *s = NULL;
+        const char *e, *service_name = name;
+
+        if (!unit_name_is_valid(name, UNIT_NAME_ANY))
+                return false;
+
+        e = endswith(name, ".service");
+        if (!e)
+                return false;
+
+        /* If it's a template or instance, get the prefix as a service name. */
+        if (unit_name_is_valid(name, UNIT_NAME_INSTANCE|UNIT_NAME_TEMPLATE)) {
+                assert_se(unit_name_to_prefix(name, &prefix) == 0);
+                assert_se(s = strjoin(prefix, ".service"));
```

Comment:
Hmm, I'd be a bit wary of using these with assert_se, otherwise we could have pid1 bail when close to OOM. Just return false here.